src/shared/settingsConfiguration.ts

```diff
@@ -5,231 +5,590 @@
+            try {
+                const value = this.config.get(key, defaultValue)
+
+                if (!descriptor) {
+                    return value as Inner[K]
+                }
+
+                if (descriptor?.[key] === undefined) {
+                    throw new Error('No type descriptor found')
+                }
+
+                return cast<Inner[K]>(value, descriptor[key])
+            } catch (error) {
+                throw new Error(`Failed to parse setting: ${error}`)
             }
-            // Caller is expected to log in the rare case that a nonexistent
-            // setting is noteworthy.
-            return undefined
         }
 
-        const actualType = typeof val
-        if (actualType !== type) {
-            const msg = `settings: getSetting(): setting "${key}": got ${actualType}, expected ${type}`
-            if (opt.silent === undefined || opt.silent !== 'yes') {
-                throw Error(`AWS Toolkit: ${msg}`)
+        public async update<K extends keyof Inner>(key: K & string, value: Inner[K]) {
+            try {
+                await this.config.update(key, value)
+
+                return true
+            } catch (error) {
+                this.log(`failed to update field "${key}": ${error}`)
+
+                return false
             }
-            if (opt.logging === undefined || opt.logging) {
-                this.log.error(msg)
+        }
+
+        public async delete(key: keyof Inner & string) {
+            try {
+                await this.config.update(key, undefined)
+
+                return true
+            } catch (error) {
+                this.log(`failed to delete field "${key}": ${error}`)
+
+                return false
+            }
+        }
+
+        public async reset() {
+            try {
+                return await this.config.reset()
+            } catch (error) {
+                this.log(`failed to reset settings: ${error}`)
             }
-            return undefined
         }
-        return val
+
+        public dispose() {
+            return vscode.Disposable.from(...this.disposables).dispose()
+        }
     }
+}
+
+/**
+ * A stricter form of {@link vscode.Memento} with type-safe semantics.
+ */
+export interface TypedSettings<T extends Record<string, any>> extends Omit<Settings, 'get' | 'update'> {
+    /**
+     * Gets the value stored at `key`.
+     *
+     * This will always return the expected type, otherwise an error is thrown. Whether a
+     * missing key is considered valid or not is dependent on the type specified.
+     */
+    get<K extends keyof T>(key: K, defaultValue?: T[K]): T[K]
 
     /**
-     * Sets a config value.
+     * Updates the value stored at `key`.
      *
-     * Writing to the (VSCode) config store may fail if the user does not have
-     * write permissions, or if some requirement is not met.  For example, the
-     * `vscode.ConfigurationTarget.Workspace` scope requires a workspace.
```

Comment:
These notes were here because the behavior is a bit surprising and definitely needs to be considered when we work with settings.

Reply:
Forgot to add these notes back in. Sometimes I get a bit overzealous with removing comments.